src/sources/docker.rs

```diff
@@ -0,0 +1,888 @@
+    out: Sender<Event>,
+    /// End through which event stream futures send ContainerLogInfo to main future
+    main_send: UnboundedSender<ContainerLogInfo>,
+}
+
+impl EventStreamBuilder {
+    fn new(
+        core: DockerSourceCore,
+        out: Sender<Event>,
+        main_send: UnboundedSender<ContainerLogInfo>,
+    ) -> Self {
+        EventStreamBuilder {
+            core,
+            out,
+            main_send,
+        }
+    }
+
+    /// Constructs and runs event stream
+    fn start(&self, id: String) -> ContainerState {
+        let mut state = ContainerState::new(id, self.core.now_timestamp);
+        self.restart(&mut state);
+        state
+    }
+
+    /// If info is present, restarts event stream
+    fn restart(&self, container: &mut ContainerState) {
+        container
+            .take_info()
+            .map(|info| self.start_event_stream(info));
+    }
+
+    fn start_event_stream(&self, info: ContainerLogInfo) {
+        // Establish connection
+        let mut options = LogsOptions::builder();
+        options
+            .follow(true)
+            .stdout(true)
+            .stderr(true)
+            // TODO: missing feature in shiplift to use .since(info.log_since())
+            .timestamps(true);
+
+        let mut stream = self
+            .core
+            .docker
+            .containers()
+            .get(&info.id)
+            .logs(&options.build());
+        info!(
+            message = "Started listening logs on docker container",
+            id = field::display(&info.id)
+        );
+
+        // Create event streamer
+        let mut state = Some((self.main_send.clone(), info));
+        let event_stream = tokio::prelude::stream::poll_fn(move || {
+            // !Hot code: from here
+            if let Some(&mut (_, ref mut info)) = state.as_mut() {
+                // Main event loop
+                loop {
+                    return match stream.poll() {
+                        Ok(Async::Ready(Some(message))) => {
+                            if let Some(event) = info.new_event(message) {
+                                Ok(Async::Ready(Some(event)))
+                            } else {
+                                continue;
+                            }
+                            // !Hot code: to here
+                        }
+                        Ok(Async::Ready(None)) => {
+                            let (main, info) = state.take().expect("They are present here");
+                            // End of stream
+                            info!(
+                                message = "Stoped listening logs on docker container",
+                                id = field::display(&info.id)
+                            );
+                            // TODO: I am not sure that it's necessary to drive this future to completition
+                            tokio::spawn(
+                                main.send(info)
```

Comment:
This should be driven since it can apply back pressure from the inner channel. I think in this case we should not spawn but drive it as apart of this stream combinator.